kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/PortConflictDetector.java

```diff
@@ -51,50 +52,58 @@ public void validate(Collection<VirtualClusterModel> virtualClusterModelMap, Opt
                 .sorted(Comparator.comparing(VirtualClusterModel::getClusterName))
                 .forEach(virtualCluster -> {
                     var name = virtualCluster.getClusterName();
-                    var proposedSharedPorts = virtualCluster.getSharedPorts();
-                    var proposedExclusivePorts = virtualCluster.getExclusivePorts();
-
-                    if (otherExclusivePort.isPresent()) {
-                        Optional<String> otherInterface = otherExclusivePort.map(hostPort -> hostPort.host().equals("0.0.0.0") ? null : hostPort.host());
-                        var checkPorts = virtualCluster.getBindAddress().isEmpty() || otherInterface.isEmpty() || virtualCluster.getBindAddress().equals(otherInterface);
-                        if (checkPorts) {
-                            checkForConflictsWithOtherExclusivePort(otherExclusivePort.get(), name, virtualCluster, proposedExclusivePorts, BindingScope.EXCLUSIVE);
-                            checkForConflictsWithOtherExclusivePort(otherExclusivePort.get(), name, virtualCluster, proposedSharedPorts, BindingScope.SHARED);
-                        }
-                    }
 
-                    // if this virtual cluster is binding to <any>, we need to check for conflicts on the <any> interface and *all* specific interfaces,
-                    // otherwise we just check for conflicts on <any> and the specified specific interface
-                    var exclusiveCheckSet = virtualCluster.getBindAddress().isEmpty() ? inUseExclusivePorts.keySet()
-                            : Set.of(ANY_INTERFACE, virtualCluster.getBindAddress());
+                    virtualCluster.listeners()
+                            .forEach((listenerName, listener) -> doValidate(otherExclusivePort, listener, name, inUseExclusivePorts, seenVirtualClusters,
+                                    inUseSharedPorts));
+
+                });
+    }
+
+    private void doValidate(Optional<HostPort> otherExclusivePort, EndpointListener listener, String name, Map<Optional<String>, Set<Integer>> inUseExclusivePorts,
```

Comment:
should we include listener name in the error messages? if I create two listeners trying to bind the same ports I see:

```
java.lang.IllegalStateException: Configuration for virtual cluster 'demo' conflicts with configuration for virtual cluster: 'demo'.
... snip ...
Caused by: java.lang.IllegalStateException: The exclusive bind of port(s) 9192,9193,9194,9195 to <any> would conflict with existing exclusive port bindings on <any>.
```

Maybe we could mention which listener we are validating?

Could also be deferred to another PR, even nicer would be to say listener X conflicts with listener Z, but looks like a bigger change.

Reply:
> edit: I have started hacking on refactoring PortConflictDetector

Ok, as you've started on this (thank you!), I'll leave it to you to open a PR against this branch.    I'd probably just take the easy option and say which listener is being validated.  I wouldn't bother trying to report precisely what it is conflicting with.

Reply:
here's the simpler change saying which listener is being validated https://github.com/k-wall/kroxylicious/pull/54

I've also had a go at the bigger refactor https://github.com/k-wall/kroxylicious/pull/55 which in the end I think is worth it for the simplification to the PortConflictDetector.

Reply:
I merged your https://github.com/k-wall/kroxylicious/pull/54.  Thank you.